boa_engine/src/builtins/iterable/async_from_sync_iterator.rs

```diff
@@ -0,0 +1,435 @@
+            .as_async_from_sync_iterator()
+            .expect("async from sync iterator prototype must be object")
+            .sync_iterator_record
+            .clone();
+
+        // 3. Let promiseCapability be ! NewPromiseCapability(%Promise%).
+        let promise_capability = PromiseCapability::new(
+            &context
+                .intrinsics()
+                .constructors()
+                .promise()
+                .constructor()
+                .into(),
+            context,
+        )
+        .expect("cannot fail with promise constructor");
+
+        // 5. If value is present, then
+        // a. Let result be Completion(IteratorNext(syncIteratorRecord, value)).
+        // 6. Else,
+        // a. Let result be Completion(IteratorNext(syncIteratorRecord)).
+        let result = sync_iterator_record.next(args.get(0).cloned(), context);
+
+        // 7. IfAbruptRejectPromise(result, promiseCapability).
+        if_abrupt_reject_promise!(result, promise_capability, context);
+
+        // 8. Return AsyncFromSyncIteratorContinuation(result, promiseCapability).
+        Self::continuation(&result, &promise_capability, context)
+    }
+
+    /// `%AsyncFromSyncIteratorPrototype%.return ( [ value ] )`
+    ///
+    /// More information:
+    ///  - [ECMA reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-%asyncfromsynciteratorprototype%.return
+    fn r#return(this: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
+        // 1. Let O be the this value.
+        // 2. Assert: O is an Object that has a [[SyncIteratorRecord]] internal slot.
+        // 4. Let syncIterator be O.[[SyncIteratorRecord]].[[Iterator]].
+        let sync_iterator = this
+            .as_object()
+            .expect("async from sync iterator prototype must be object")
+            .borrow()
+            .as_async_from_sync_iterator()
+            .expect("async from sync iterator prototype must be object")
+            .sync_iterator_record
+            .iterator()
+            .clone();
+
+        // 3. Let promiseCapability be ! NewPromiseCapability(%Promise%).
+        let promise_capability = PromiseCapability::new(
+            &context
+                .intrinsics()
+                .constructors()
+                .promise()
+                .constructor()
+                .into(),
+            context,
+        )
+        .expect("cannot fail with promise constructor");
+
+        // 5. Let return be Completion(GetMethod(syncIterator, "return")).
+        let r#return = sync_iterator.get_method("return", context);
+
+        // 6. IfAbruptRejectPromise(return, promiseCapability).
+        if_abrupt_reject_promise!(r#return, promise_capability, context);
+
+        let result = match (r#return, args.get(0)) {
+            // 7. If return is undefined, then
+            (None, _) => {
+                // a. Let iterResult be CreateIterResultObject(value, true).
+                let iterator_result =
```

Comment:
This might better adhere to the spec if it would be called `iter_result`, but I think this name is also pretty clear.